conda_build_prepare/conda_cmds.py

```diff
@@ -27,11 +111,254 @@ def create_env(package_dir):
     for lc in reversed(get_local_channels()):
         data['channels'].insert(0, lc)
 
-    prefix = run_conda("create", "--name", package, "--strict-channel-priority")
+    prefix = run("create", "--name", package, "--strict-channel-priority")
     env_condarc = os.path.join(prefix, 'condarc')
     with open(env_condarc, "w") as f:
         yaml.safe_dump(data, f)
 
     print(open(env_condarc).read())
 
 
+def _prepare_single_source(git_repos_dir, src):
+    if 'git_url' not in src.keys() or 'git_rev' not in src.keys():
+        # Not a git source; won't be prepared and is not modified
+        return src
+    else:
+        # Clone, checkout the repository and set local path as git_url
+        src_path = git_clone(src['git_url'], git_repos_dir)
+        git_checkout(src_path, src['git_rev'])
+        git_clone_relative_submodules(src_path, src['git_url'])
+        src['git_url'] = os.path.abspath(src_path)
+        return src
+
+def _add_extra_tags_if_exist(package_dir, repo_path):
+    assert os.path.exists(package_dir)
+    extra_tags_path = os.path.join(package_dir, 'extra.tags')
+
+    if not os.path.exists(extra_tags_path):
+        return
+    else:
+        print()
+        print('Adding tags from extra.tags file...')
+        with open(extra_tags_path, 'r') as extra_tags_file:
+            for line in extra_tags_file.readlines():
+                try:
+                    tag, sha = line.replace('\n', '').split()
+                    git_add_tag(repo_path, tag, sha, temp_user=True)
+                except ValueError:
+                    print('Malformed line in extra.tags:')
+                    print('\t"{}"'.format(line.replace('\n', '')))
+                    print()
+                    continue
+                except subprocess.CalledProcessError:
+                    print('Git had problems adding extra.tags line:')
+                    print('\t"{}"'.format(line.replace('\n', '')))
+                    print()
+                    continue
+        print()
+
+def _call_conda_cmd_in_env(cmd_string, env_path):
+    assert os.path.isdir(env_path), env_path
+
+    return subprocess.check_output(f'conda run -p {env_path} {cmd_string}'
```

Comment:
You probably don't want to use `subprocess.check_output` with arbitrary conda commands as there is complicated blocking issues around pipes.

You should instead create temporary files and use `stdout=<file>`. There is a bunch of complexity around `close_fds=True` and `pass_fds=()`...